game-core/src/main/java/games/strategy/triplea/ai/pro/util/MoveBatcher.java

```diff
@@ -0,0 +1,164 @@
+package games.strategy.triplea.ai.pro.util;
+
+import games.strategy.engine.data.Route;
+import games.strategy.engine.data.Unit;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * MoveBatcher implements an algorithm that batches AI moves together, reducing the number of
+ * distinct moves the AI does (and thus the visual delay) by combining moves together.
+ */
+public class MoveBatcher {
+  private static class Move {
+    private final ArrayList<Unit> units;
+    private final Route route;
+    private final HashMap<Unit, Unit> unitsToTransports;
+
+    private Move(
+        final ArrayList<Unit> units,
+        final Route route,
+        final HashMap<Unit, Unit> unitsToTransports) {
+      this.units = units;
+      this.route = route;
+      this.unitsToTransports = unitsToTransports;
+    }
+
+    Move(final ArrayList<Unit> units, final Route route) {
+      this(units, route, null);
+    }
+
+    Move(final Unit unit, final Route route, final Unit transportToLoad) {
+      this(
+          mutableSingletonList(unit),
+          route,
+          new HashMap<Unit, Unit>(Map.of(unit, transportToLoad)));
+    }
+
+    private static ArrayList<Unit> mutableSingletonList(final Unit unit) {
+      return new ArrayList<>(Collections.singletonList(unit));
+    }
+
+    boolean isTransportLoad() {
+      return this.unitsToTransports != null;
+    }
+
+    boolean canMergeWith(final Move other) {
+      return other.isTransportLoad() == isTransportLoad() && route.equals(other.route);
+    }
+
+    boolean mergeWith(final Move other) {
+      if (canMergeWith(other)) {
+        // Merge units and transports.
+        units.addAll(other.units);
+        if (isTransportLoad()) {
+          unitsToTransports.putAll(other.unitsToTransports);
+        }
+        return true;
+      }
+      return false;
+    }
+
+    void addTo(
+        final List<Collection<Unit>> moveUnits,
+        final List<Route> moveRoutes,
+        final List<Map<Unit, Unit>> unitsToTransports) {
+      moveUnits.add(units);
```

Comment:
This is jumping out to me, sorry I don't quite have the time to look at this in detail. Is there a way though where we can invert the data dependency relationship? To explain, we are making exclusive use of 'out' parameters, I think it's potentially a violation of encapsulation. Any caller to this code must know the state of the current object to then know the state of the 'out' parameters. That is potentially not going to be good because whenever reading the caller code, you need to mentally keep track, or a tally, of the state of this object to understand that code.

I'll try to take a deeper look tomorrow, just wanted to get this conversation started and to highlight this design area for further review.

Reply:
This is just for interfacing with the existing way moves are handled, see ProMoveUtils.java and IMoveDelegate.java.

I agree that this is messy and I think we should just refactor things so there's a single Move object that's used - instead of these weird parallel lists. But this change is already large enough and it's out of scope for it.

Reply:
Okay, that sounds fair enough. Would you mind commenting the code to document that? "TODO: #5416 follow-up Refactor to avoid the parallel lists, the side effects are to support existing code" (super-paraphrased).

I don't mind us doing what's needed to get things done, it's not all going to be pretty, but we can perhaps mitigate it a bit and have pity on the next reader of the code with some call-out comments of where we know things are not ideal and would ideally be fixed up later.